Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+ */
+
+/// A gRPC client.
+///
+/// A ``GRPCClient`` communicates to a server via a given ``ClientTransport``.
+/// You can start RPCs to the server by calling the corresponding method:
+/// - ``unary(request:descriptor:serializer:deserializer:handler:)``
+/// - ``clientStreaming(request:descriptor:serializer:deserializer:handler:)``
+/// - ``serverStreaming(request:descriptor:serializer:deserializer:handler:)``
+/// - ``bidirectionalStreaming(request:descriptor:serializer:deserializer:handler:)``
+///
+/// You can set ``MethodConfiguration``s on this client to override whatever configurations have been
+/// set on the given transport.
+/// You can also use ``ClientInterceptor``s to implement cross-cutting logic which apply to all
+/// RPCs. Example uses of interceptors include authentication and logging.
+///
+/// ## Creating and configuring a client
+///
+/// The following example demonstrates how to create and configure a server.
+///
+/// ```swift
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessClientTransport()
+/// let client = GRPCClient(transport: inProcessTransport)
+///
+/// // Create and add some interceptors.
+/// client.interceptors.add(StatsRecordingServerInterceptors())
+///
+/// // Create and add some method configurations.
+/// let defaultConfiguration = MethodConfiguration(
+///     executionPolicy: ...,
+///     timeout: ...
+/// )
+/// let registry = MethodConfigurationRegistry(defaultConfiguration: defaultConfiguration)
+/// client.methodConfigurationOverrides = registry
+/// ```
+///
+/// ## Starting and stopping the client
+///
+/// Once you have configured the client, call ``run()`` to start it. Calling ``run()`` connects to the given
+/// transport.
+///
+/// ```swift
+/// // Start running the client.
+/// try await client.run()
+/// ```
+///
+/// The ``run()`` method won't return until the client has finished handling all requests. You can
+/// signal to the client that it should stop creating new request streams by calling ``close()``.
+/// This gives the client enough time to drain any requests already in flight. To stop the client more abruptly
+/// you can cancel the task running your client. If your application requires additional resources
+/// that need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public final class GRPCClient: Sendable {
+  /// A collection of ``ClientInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request sent from the client to
+  /// the server will first be intercepted by the first added interceptor followed by the second, and so on.
+  /// For responses from the server, they'll be applied in the opposite order.
+  public var interceptors: Interceptors {
```

Comment:
What happens if I change interceptors while a request is currently making its way through the interceptor pipeline?

Reply:
You can't modify interceptors once the Client is running (and you cannot execute RPCs unless the Client is running either) so this cannot happen.

Reply:
Nothing. There's a state check _and_ we grab the array of interceptors before executing the request.

Reply:
I see but I also find it a bit weird. Like you are calling `interceptors = []` after calling run and that call just gets dropped and nothing happens. Why don't we just make them a part of the init and not allow modifying them afterwards?